src/Authoring/WinRT.SourceGenerator/Analyzers/ComImportInterfaceAnalyzer.cs

```diff
@@ -0,0 +1,116 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#if ROSLYN_4_12_0_OR_GREATER
+
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Operations;
+using WinRT.SourceGenerator;
+
+#nullable enable
+
+namespace Generator;
+
+/// <summary>
+/// A diagnostic analyzer to warn for casts to <see cref="System.Runtime.InteropServices.ComImportAttribute"/> interfaces.
+/// </summary>
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class ComImportInterfaceAnalyzer : DiagnosticAnalyzer
+{
+    /// <inheritdoc/>
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = [WinRTRules.ComImportInterfaceCast];
+
+    /// <inheritdoc/>
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+        context.EnableConcurrentExecution();
+
+        context.RegisterCompilationStartAction(static context =>
+        {
+            // We only need to emit warnings if CsWinRT is in 'auto' mode (same as the collection expressions analyzer)
+            if (!GeneratorExecutionContextHelper.IsCsWinRTAotOptimizerInAutoMode(context.Options.AnalyzerConfigOptionsProvider, context.Compilation))
```

Comment:
I feel like you should keep the GetEnableAotAnalyzer check here still as someone may want to have auto mode enabled here while still not publishing for AOT yet.

Reply:
But isn't the point that if the AOT optimizer we should still emit all possible AOT warnings? I mean we do the same for the collection expression analyzer, for instance, even though that code would technically work if you don't set `PublishAot`. I'm happy to change this if you prefer, I guess I'm just a bit confused by the exact matrix of possible combinations for all the available AOT analyzer settings we have 😅

Reply:
I see the collection expression analyzer a bit different as trimming scenarios can also be impacted by it given we won't generate the vtables and interfaces can get trimmed and so on.  Whereas this, I can see this working fine on trimming while being in auto mode.  So far, all our aot optimizer scenarios have been also useful with just trimming enabled.  I think this is the first one that is going to differ in that behavior.

Reply:
I see, fixed! 🙂